### PR TITLE
User/ryano/azure integration

### PR DESCRIPTION
**Pull request by ryanorbaker**
1. Changes to add Azure configuration.
2. Changes to allow local configuration file called: appsettings.Development.json.   Values in this file will NOT be added to the git repo, and supersede the values in the default appsettings.json file.

---
This change is [Reviewable](https://reviewable.io/reviews/healthcatalyst/fabric.identity/251)

### CODE REVIEW

File: Fabric.Identity.API/Configuration/IAppConfiguration.cs [hunk 1]
```diff
@@ -4,6 +4,8 @@
 
 namespace Fabric.Identity.API.Configuration
 {
+    using System.Collections.Generic;
```
Comment (michaelvidal24) [inline Fabric.Identity.API/Configuration/IAppConfiguration.cs]: Do we need this using?

File: Fabric.Identity.API/Quickstart/Account/AccountController.cs [hunk 2]
```diff
@@ -237,9 +242,27 @@ public async Task<IActionResult> ExternalLoginCallback(string returnUrl)
...
+
             if (userIdClaim == null)
             {
-                throw new Exception("Unknown userid");
+                throw new MissingUserClaimException(ExceptionMessageResources.MissingUserClaimMessage);
+            }
+
+            var schemaItem = info.Properties.Items.FirstOrDefault(i => i.Key == "scheme");
+
+            if (_appConfiguration.AzureAuthenticationEnabled && schemaItem.Value == FabricIdentityConstants.AuthenticationSchemes.Azure)
```
Comment (michaelvidal24) [inline Fabric.Identity.API/Quickstart/Account/AccountController.cs]: Do we need to check schemaItem for null here?

File: Fabric.Identity.API/Services/UserProfileService.cs [hunk 3]
```diff
@@ -54,7 +60,16 @@ public async Task IsActiveAsync(IsActiveContext context)
             var sub = context.Subject.GetSubjectId();
             _logger.Debug($"found sub from IsActiveContext: {sub}");
             var user = await _userStore.FindBySubjectIdAsync(sub);
+
             context.IsActive = user != null;
+
+            if (user != null && _appConfig.AzureAuthenticationEnabled)
```
Comment (michaelvidal24) [inline Fabric.Identity.API/Services/UserProfileService.cs]: Do we also have to check the scheme that the user logged in with? If we have Windows auth enabled and Azure auth enabled, we might set the user to inactive if they are a windows user and we don't have an issuer for them.

File: Fabric.Identity.API/Startup.cs [hunk 4]
```diff
@@ -208,6 +209,11 @@ public void ConfigureServices(IServiceCollection services)
             app.UseCors(FabricIdentityConstants.FabricCorsPolicyName);
 
             app.UseIdentityServer();
+            if (_appConfig.AzureAuthenticationEnabled)
```
Comment (michaelvidal24) [inline Fabric.Identity.API/Startup.cs]: Could we encapsulate this check in the UseAzureIdentityProvider function? We are passing in the config there anyway - might help to keep this already messy class a little cleaner.

File: Fabric.Identity.API/appsettings.json [hunk 5]
```diff
@@ -42,6 +43,7 @@
   "ExternalIdProviderSettings": {
     "ExternalIdProviders": []
   },
+  "AzureActiveDirectorySettings": {},
```
Comment (michaelvidal24) [inline Fabric.Identity.API/appsettings.json]: Do we want to add some default settings here, so we can add comments for what the settings are for?

File: Fabric.Identity.API/Configuration/IdentityConfigurationProvider.cs [hunk 6]
```diff
@@ -4,24 +4,26 @@
...
-            return BuildAppConfiguration(basePath);
+            return BuildAppConfiguration(basePath, string.Empty);
         }
 
-        private IAppConfiguration BuildAppConfiguration(string baseBath)
+        private IAppConfiguration BuildAppConfiguration(string baseBath, string environmentName)
         {
                 var config = new ConfigurationBuilder()
                     .AddJsonFile("appsettings.json")
+                    .AddJsonFile($"appsettings.{environmentName}.json", true)
```
Comment (michaelvidal24) [inline Fabric.Identity.API/Configuration/IdentityConfigurationProvider.cs]: I Like this addition!

File: Fabric.Identity.API/Configuration/IAppConfiguration.cs (hunk 1 above)
Comment (ryanorbaker) [reply Fabric.Identity.API/Configuration/IAppConfiguration.cs]: nope, it's a hangover.

File: Fabric.Identity.API/Quickstart/Account/AccountController.cs (hunk 2 above)
Comment (ryanorbaker) [reply Fabric.Identity.API/Quickstart/Account/AccountController.cs]: nope.  schemaItem is actually a key/value pair.  A dictionary returns a "","" pair if first or default is not found.

File: Fabric.Identity.API/Startup.cs (hunk 4 above)
Comment (ryanorbaker) [reply Fabric.Identity.API/Startup.cs]: ha, yes.   I was thinking the startup was big to make it more straightforward to read :-)

File: Fabric.Identity.API/Services/UserProfileService.cs (hunk 3 above)
Comment (ryanorbaker) [reply Fabric.Identity.API/Services/UserProfileService.cs]: good point.  I wonder how this is working!

File: Fabric.Identity.API/appsettings.json (hunk 5 above)
Comment (ryanorbaker) [reply Fabric.Identity.API/appsettings.json]: done

File: Fabric.Identity.API/Services/UserProfileService.cs (hunk 3 above)
Comment (ryanorbaker) [reply Fabric.Identity.API/Services/UserProfileService.cs]: updated

File: Fabric.Identity.API/Extensions/ApplicationBuilderExtensions.cs [hunk 7]
```diff
@@ -5,14 +5,56 @@
...
 
 namespace Fabric.Identity.API.Extensions
 {
+    using Microsoft.IdentityModel.Protocols.OpenIdConnect;
+
     public static class ApplicationBuilderExtensions
     {
+
+
+        public static IApplicationBuilder UseAzureIdentityProviderIfAccplicable(this IApplicationBuilder builder,
```
Comment (michaelvidal24) [inline Fabric.Identity.API/Extensions/ApplicationBuilderExtensions.cs]: Check spelling here: UseAzureIdentityProviderIfApplicable.

File: Fabric.Identity.API/Fabric.Identity.API.csproj [hunk 8]
```diff
@@ -32,6 +32,10 @@
     <None Remove="Persistence\SqlServer\Entities\**" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Content Remove="appsettings.Development.json" />
```
Comment (hc-jordenlowe) [inline Fabric.Identity.API/Fabric.Identity.API.csproj]: is this file described somewhere?  Is that the documentation you are creating?

Comment (ryanorbaker) [reply Fabric.Identity.API/Fabric.Identity.API.csproj]: this is a copy file of the appsettings.json but it take precedence.   so you can have bogus values in appsettings and then create this file locally with your real secrets.

Comment (hc-jordenlowe) [reply Fabric.Identity.API/Fabric.Identity.API.csproj]: ok

**Pull request merged**